lxd/ovn.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/canonical/lxd/lxd/state"
+	"github.com/canonical/lxd/shared/api"
+)
+
+type OVNLogMonitor struct {
+	stateFunc func() *state.State
+	conn      net.PacketConn
+}
+
+// NewOVNLogMonitor returns a new OVNLogMonitor.
+func NewOVNLogMonitor(fn func() *state.State) *OVNLogMonitor {
+	if fn == nil {
+		panic("Invalid state function")
+	}
+
+	return &OVNLogMonitor{
+		stateFunc: fn,
+	}
+}
+
+// Start starts the log monitor. It listens for log messages using UDP, and sends them to the events API.
+func (l *OVNLogMonitor) Start(ctx context.Context, address string) (context.CancelFunc, error) {
+	var listenConfig net.ListenConfig
+	var err error
+
+	messages := make(chan string)
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	l.conn, err = listenConfig.ListenPacket(ctx, "udp", address)
+	if err != nil {
+		cancelFunc()
+		return nil, err
+	}
+
+	// This is the goroutine that processes the log messages. It also closes the connection if the context is done (this will terminate the second goroutine below as `ReadFrom` will receive an error).
+	go func() {
+		var applicationName string
+		var sequenceNumber string
+		var moduleName string
+		var logLevel string
+		var message string
+
+		for {
+			select {
+			case <-ctx.Done():
+				// This closes the connection and makes `ReadFrom` return an error (see goroutine below).
+				l.conn.Close()
+				return
+			case msg := <-messages:
+				// Acceptable formats:
+				// - <29> ovs|00017|rconn|INFO|unix:/var/run/openvswitch/br-int.mgmt: connected"
+				// - <29> ovs|ovn-controller|00017|rconn|INFO|unix:/var/run/openvswitch/br-int.mgmt: connected"
+				// The first field can be ignored as that information is relevant to syslogd.
+				fields := strings.SplitN(msg, "|", 6)
+
+				if len(fields) == 5 {
+					applicationName = ""
+				} else if len(fields) == 6 {
+					applicationName = fields[1]
+				} else {
+					continue
+				}
+
+				sequenceNumber = fields[len(fields)-4]
+				moduleName = fields[len(fields)-3]
+				logLevel = strings.ToLower(fields[len(fields)-2])
+				message = fields[len(fields)-1]
+
+				event := api.EventLogging{
+					Level:   logLevel,
+					Message: message,
+					Context: map[string]string{
+						"application": applicationName,
+						"module":      moduleName,
+						"sequence":    sequenceNumber,
+					},
+				}
+
+				if applicationName != "" {
+					event.Context["application"] = applicationName
+				}
+
+				err = l.stateFunc().Events.Send("", api.EventTypeOVN, event)
```

Comment:
As we only use stateFunc here to get Events, shall we just pass in d.events itself in `NewOVNLogMonitor` above, like we do for `events.NewInternalListener(d.shutdownCtx, d.events)`. This would save repeated calls to `stateFunc`.